mcs/class/System/System.Net/ServicePointScheduler.cs

```diff
@@ -245,9 +245,14 @@ void RunSchedulerIteration ()
 				Debug ($"ITERATION #1: {repeat} {groups != null}");
 
 				if (groups != null) {
-					foreach (var group in groups) {
-						Debug ($"ITERATION #2: group={group.Value.ID}");
-						repeat |= SchedulerIteration (group.Value);
+					var keys = new string[groups.Count];
```

Comment:
Doesn't this just hide the race bellow where you do `groups.Remove (groupName);` without a lock?

Reply:
There is no "race condition" here in the traditional sense and no locks or concurrent dictionaries are going to help you here.

What's happening here is that `SchedulerIteration()` might actually remove the current group from the list when it fails to establish a connection.

Prior to the commit that I just pushed on top of #7358, there as also a separate issue where arbitrary user code might get executed from within `SchedulerIteration()` (on our thread, while holding the lock) as part of a `TaskCompletionSource`s continuation.

This has now been fixed by replacing all uses of `TaskCompletionSource` with `WebCompletionSource`, which uses `TaskCreationOptions.RunContinuationsAsynchronously`.